lib/matrix.ex

```diff
@@ -0,0 +1,332 @@
+
+  ##  Example:
+
+      iex> {:ok, mat} = Matrix.new(4)
+      iex> Matrix.update_row(mat, [1, 2, 3], {3, 1})
+      {:ok,
+        [
+          [0, 0, 0, 0],
+          [0, 0, 0, 0],
+          [0, 0, 0, 0],
+          [0, 1, 2, 3]
+        ]
+      }
+
+  """
+  @spec update_row(matrix, element, index) :: Result.t(String.t(), matrix)
+  def update_row(matrix, submatrix, index) do
+    update(matrix, [submatrix], index)
+  end
+
+  @doc """
+  Updates column in the matrix by given a column vector (list) of numbers.
+  The column and rows which you want to change is given by tuple `{row, col}`.
+  Both values are non negative integers.
+
+  Returns result, it means either tuple of {:ok, matrix} or {:error, "msg"}.
+
+  ##  Example:
+
+      iex> {:ok, mat} = Matrix.new(4)
+      iex> Matrix.update_col(mat, [[1], [2], [3]], {0, 1})
+      {:ok,
+        [
+          [0, 1, 0, 0],
+          [0, 2, 0, 0],
+          [0, 3, 0, 0],
+          [0, 0, 0, 0]
+        ]
+      }
+
+  """
+  @spec update_col(matrix, element, index) :: Result.t(String.t(), matrix)
+  def update_col(matrix, submatrix, index) do
+    update(matrix, submatrix, index)
+  end
+
+  @doc """
+  Updates the matrix by given a submatrices. The positions (or locations) of these
+  submatrices are given by list of indices. Index of the individual submatrices is
+  tuple of two numbers. These two numbers are row and column of matrix where the
+  submatrices will be located. All submatrices must have same size (dimension).
+
+  Returns result, it means either tuple of {:ok, matrix} or {:error, "msg"}.
+
+  ##  Example:
+
+      iex> mat = Matrix.new(5)
+      iex> sub_mat = Matrix.new(2,1)
+      iex> positions = [{0,0}, {3, 3}]
+      iex> Matrix.update_map(mat, sub_mat, positions)
+      {:ok,
+        [
+          [1, 1, 0, 0, 0],
+          [1, 1, 0, 0, 0],
+          [0, 0, 0, 0, 0],
+          [0, 0, 0, 1, 1],
+          [0, 0, 0, 1, 1]
+        ]
+      }
+
+  """
+  @spec update_map(matrix, element, list(index)) :: Result.t(String.t(), matrix)
+  def update_map(matrix, submatrix, positions) do
+    {row, col, check_size} = and_then2(matrix, submatrix, &check_size(&1, &2, positions))
+
+    if check_size do
+      Enum.reduce(positions, matrix, fn position, acc ->
+        and_then2(
+          acc,
+          submatrix,
+          &update(&1, &2, position)
+        )
+      end)
+    else
+      Result.error("Bad value of position {#{row}, #{col}}. Submatrix is out of matrix!")
+    end
+  end
+
+  @doc """
+  Transpose of matrix.
+
+  Returns result, it means either tuple of {:ok, matrix} or {:error, "msg"}.
+
+  ##  Example:
+
+      iex> mat = {:ok, [[1,2,3], [4,5,6], [7,8,9]]}
+      iex> mat |> Result.map(&Matrix.transpose(&1))
+      {:ok,
+        [
+          [1, 4, 7],
+          [2, 5, 8],
+          [3, 6, 9]
+        ]
+      }
+
+  """
+  @spec transpose(matrix) :: Result.t(String.t(), matrix)
```

Comment:
Tato fce ale nevraci result, ale jen matrix, neni volana v `Result.and_then/2`, ale `Result.map/2`, kde `Result.and_then/2` ma typ `Result.t(a, b) -> (b -> Result.t(a, c)) -> Result.t(a, c)`, ale `Result.map/2` ma typ `Result.t(a, b) -> (b -> c) -> Result.t(a, c)`. Je to videt i z toho prikladu a dokonce i *dialyzer` s tim ma problem.